source/protocol/lsp-types.adb

```diff
@@ -166,6 +170,113 @@ package body LSP.Types is
       return Ada.Strings.UTF_Encoding.Wide_Strings.Encode (Wide);
    end To_UTF_8_String;
 
+   -------------------------------
+   -- To_UTF_8_Unbounded_String --
+   -------------------------------
+
+   function To_UTF_8_Unbounded_String
+     (Value : LSP_String) return GNATCOLL.JSON.UTF8_Unbounded_String
+   is
+      function To_Unsigned_16 is new
+        Ada.Unchecked_Conversion (Wide_Character, Interfaces.Unsigned_16);
+
+      type Wide_String_Access is access Wide_String;
+      procedure Unchecked_Free is new Ada.Unchecked_Deallocation
+        (Wide_String, Wide_String_Access);
+
+      type String_Access is access String;
+      procedure Unchecked_Free is new Ada.Unchecked_Deallocation
+        (String, String_Access);
+
+      function Encode
+        (Item       : Wide_String_Access;
+         Result     : String_Access) return Natural;
+      --  Put in Result the contents of Item in UTF-8.
+      --  Result must be preallocated to (1 .. Item'Length * 3.
+      --  The Length of the actual result is returned.
+      --  ??? This is a copy of Ada.Strings.UTF_Encoding.Wide_Strings.Encode
+      --  which does not use strings allocated on the stack.
+
+      ------------
+      -- Encode --
+      ------------
+
+      function Encode
+        (Item       : Wide_String_Access;
+         Result     : String_Access) return Natural
+      is
+         use Interfaces;
+
+         Len : Natural;
+         --  Number of output codes stored in Result
+
+         C : Unsigned_16;
+         --  Single input character
+
+         procedure Store (C : Unsigned_16);
+         pragma Inline (Store);
+         --  Store one output code, C is in the range 0 .. 255
+
+         -----------
+         -- Store --
+         -----------
+
+         procedure Store (C : Unsigned_16) is
+         begin
+            Len := Len + 1;
+            Result (Len) := Character'Val (C);
```

Comment:
Can we use this:
```
Ada.Strings.Unbounded.Append (Result, Character'Val (C));
```
Average complexity of Append is O(1), so not worse, but we avoid explicit `new` and `Free` in the code.

This way we can do even better: just iterate over Value taking slices of some reasonable length (say 512 characters) Encode them and append to Result. Like this:

```
Ada.Strings.Unbounded.Append (Result, Encode (Value (J .. J + 511)));
```

So we don't need copy/pasting of Encode.
What do you think?

Reply:
very good idea, I will try!

Reply:
Done!